janusgraph-lucene/src/main/java/org/janusgraph/diskstorage/lucene/LuceneIndex.java

```diff
@@ -561,10 +561,14 @@ private void tokenize(SearchParams params, final Mapping mapping, final LuceneCu
                 params.addQuery(new PrefixQuery(term), BooleanClause.Occur.MUST);
             } else throw new IllegalArgumentException("LuceneIndex does not support this predicate with 1 token : " + janusgraphPredicate);
         } else {
-            // at the moment, this is only walked for EQUAL and Text.CONTAINS (String and Text mappings)
+            // at the moment, this is only walked for EQUAL, NOT_EQUAL and Text.CONTAINS (String and Text mappings)
             final BooleanQuery.Builder q = new BooleanQuery.Builder();
             for (final String term : terms) {
-                q.add(new TermQuery(new Term(key, term)), BooleanClause.Occur.MUST);
+                if (janusgraphPredicate == Cmp.NOT_EQUAL) {
```

Comment:
Shouldn't we also change the logic of `Cmp.NOT_EQUAL` if there is only 1 term (line 549)?
I didn't check the logic but when there is 1 term only it uses next query:
```
final BooleanQuery.Builder q = new BooleanQuery.Builder();
q.add(new MatchAllDocsQuery(), BooleanClause.Occur.MUST);
q.add(new TermQuery(new Term(key, terms.get(0))), BooleanClause.Occur.MUST_NOT);
params.addQuery(q.build(), BooleanClause.Occur.MUST);
``` 
And when there is multiple terms it uses:
```
q.add(new TermQuery(new Term(key, term)), BooleanClause.Occur.MUST_NOT);
```

Reply:
I'm restore MatchAll logic if NOT_EQUAL